### PR TITLE
Add logic to get selection handler, with tests

### PR DESCRIPTION
**Pull request by RoySegall**

### CODE REVIEW

**Comment by RoySegall** [comment]
@damiankloip I would like to start test the controller. Do you know of any documentation for phpunit that checks entity reference fields? I don't want to go into simple test for this one.

**Comment by damiankloip** [comment]
@RoySegall Do you mean your widget classes? If so, you might want to look at `EntityUnitTestBase`, however this is extending `KernelTestBase`, not Just `UnitTestCase`. But now there is a new `KernelTests\KernelTestBase` which I think is the way to go. See `EntityReferenceSelectionReferenceableTest` as a good example of that.

**Comment by RoySegall** [comment]
@damiankloip `EntityReferenceSelectionReferenceableTest` is a good jumpstart. 10x.

**Comment by RoySegall** [comment]
@damiankloip Maybe you could find out the problem. When creating the audience field in the test we get and [error](https://travis-ci.org/amitaibu/og/builds/86015501#L122) 

> Drupal\Core\Config\Schema\SchemaIncompleteException: Schema errors for field.storage.node.og_group_ref with the following errors: field.storage.node.og_group_ref:settings.target_type missing schema

If i'll change [this](https://github.com/amitaibu/og/blob/8.x-1.x/src/Plugin/OgFields/AudienceField.php#L30) to `entity_reference` from `og_membership_reference` we won't get an error.

File: config/schema/og.schema.yml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 og.settings:
-  type: config_object
+  type: config_entity
```
Comment (damiankloip) [inline config/schema/og.schema.yml]: Not sure why this was changed? This `og.settings` schema is just for simple config, not a config_entity. Is this not why the tests are breaking?

File: src/Kernel/Entity/SelectionHandlerTest.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\og\Og;
+use Drupal\user\Entity\User;
+
+/**
+ * Tests entity reference selection plugins.
+ *
+ * @group og
+ */
+class SelectionHandlerTest extends KernelTestBase {
```
Comment (damiankloip) [inline src/Kernel/Entity/SelectionHandlerTest.php]: <3

File: config/schema/og.schema.yml (hunk 1 above)
Comment (RoySegall) [inline config/schema/og.schema.yml]: When changing to config entity the test did not broke at the beginning. Now the test fails for something else.

Comment (damiankloip) [inline config/schema/og.schema.yml]: As far as I know,  that's just not right :)

**Comment by RoySegall** [comment]
@damiankloip Could you have a look on my previous comment? 10x.

File: config/schema/og.schema.yml (hunk 1 above)
Comment (damiankloip) [inline config/schema/og.schema.yml]: This just moves the problem to somewhere else. This config type is not the issue.

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 3]
```diff
@@ -42,9 +41,34 @@ public function getAccount() {
...
+    return empty($this->configuration[$key]) ? $this->configuration : $this->configuration[$key];
+  }
+
+  /**
+   * Get the handler of the field.
+   *
+   * @return DefaultSelection
+   */
+  public function getHandler() {
+    return \Drupal::service('plugin.manager.entity_reference_selection')->getInstance($this->configuration);
```
Comment (RoySegall) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: HUH?

Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: Ha :) ok, that is pretty much what we want

Comment (RoySegall) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: But i should call our own plugin manager - https://github.com/amitaibu/og/pull/11/files#diff-74d260cea47b42dffdde5e684823d336R86

**Comment by RoySegall** [comment]
The only thing I have left to solve it's [this](https://travis-ci.org/amitaibu/og/builds/88609189#L123):

> Drupal\Core\Config\Schema\SchemaIncompleteException: Schema errors for field.storage.node.og_group_ref with the following errors: field.storage.node.og_group_ref:settings.target_type missing schema

File: .travis.yml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 language: php
 sudo: false
 
-php: 5.5
+php: 5.6
```
Comment (damiankloip) [inline .travis.yml]: We should test PHP 5.5 here too, as D8 minimum PHP version requirement is 5.5.

File: src/Og.php [hunk 5]
```diff
@@ -58,21 +58,21 @@ public static function createField($field_name, $entity_type, $bundle) {
...
-    // Define the view mode for the field.
-    if ($fieldViewModes = $og_field->viewModesDefinition()) {
-      $prefix = $entity_type . '.' . $bundle . '.';
-      $viewModes = \Drupal::entityManager()->getStorage('entity_view_display')->loadMultiple(array_keys($fieldViewModes));
-
-      foreach ($viewModes as $key => $viewMode) {
-        $viewMode->setComponent($field_name, $fieldViewModes[$prefix . $key])->save();
-      }
-    }
+//    $displayForm = EntityFormDisplay::load($entity_type . '.' . $bundle . '.default');
```
Comment (damiankloip) [inline src/Og.php]: Does this now mean when I create a field I will not get a placed widget?

File: src/Og.php [hunk 6]
```diff
@@ -273,4 +273,18 @@ protected static function fieldInfo($field_name = NULL) {
...
+  /**
+   * Check if the given name is an audience field.
+   *
+   * @param $field_name
+   *   The field name.
+   *
+   * @return bool
+   *   TRUE if the field name represent audience field.
+   */
+  public static function isAudienceField($field_name) {
```
Comment (damiankloip) [inline src/Og.php]: We already have a method for this that correctly checks the field type. See https://github.com/amitaibu/og/commit/3993f9a8c516e5e95678901ac966912e56465e18

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 7]
```diff
@@ -16,9 +16,8 @@
  * Provide default OG selection handler.
  *
  * @EntityReferenceSelection(
- *   id = "default:og",
+ *   id = "og:default",
```
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:26]: :+1: I was going to mention that too :)

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 8]
```diff
@@ -42,9 +41,34 @@ public function getAccount() {
    * Set the current object account.
    *
    * @param AccountInterface $account
+   *   The user object.
+   * @return $this
    */
-  public function setAccount(AccountInterface $account) {
+  public function setAccount( $account) {
```
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: Why this change? You want to set NULL? If so just change the parameter to:

``` php
public function setAccount(AccountInterface $account = NULL) {
```

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 9]
```diff
@@ -42,9 +41,34 @@ public function getAccount() {
...
+   * Get the field configuration.
+   *
+   * @param $key
+   *   Specific key from the configuration. Optional.
+   *
+   * @return array|string
+   *   Value of a configuration or all the configurations.
+   */
+  public function getConfiguration($key = NULL) {
+    return empty($this->configuration[$key]) ? $this->configuration : $this->configuration[$key];
```
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: IMO this is not a good thing to do, if you pass in a key that doesn't exist, you will get all configuration. That's a bit of a WTF. We should check isset($key) first before returning all configuration.

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 10]
```diff
@@ -42,9 +41,34 @@ public function getAccount() {
...
+  }
+
+  /**
+   * Get the handler of the field.
+   *
+   * @return DefaultSelection
+   */
+  public function getHandler() {
+    return \Drupal::service('og.selection_manager')->getInstance($this->configuration);
```
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: We should inject this dependency, along with the current user too. Otherwise testing becomes pretty painful. We can do that in a follow up though if you like?

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 11]
```diff
@@ -42,9 +41,34 @@ public function getAccount() {
...
+  public function getConfiguration($key = NULL) {
+    return empty($this->configuration[$key]) ? $this->configuration : $this->configuration[$key];
+  }
+
+  /**
+   * Get the handler of the field.
+   *
+   * @return DefaultSelection
+   */
+  public function getHandler() {
```
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: Maybe we should name this a bit better. `getBaseSelectionHandler` or something?

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 12]
```diff
@@ -78,7 +104,7 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
 
     $ids = [];
 
-    if ($this->configuration['handler_settings']['field_mode'] == 'admin') {
+    if (empty($this->configuration['handler_settings']['field_mode'])) {
```
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: Checking for empty() like this on a setting is asking for more bugs, I think we should keep explicit checking for types, if it's not set for whatever reason, we want to know about it. Can you look at the changes in https://github.com/amitaibu/og/pull/21, these changes make the selection handler work for me. This being the case:

I am a member of group A and B, but NOT group C. I see groups A and B in the regular selection autocomplete, and group C only in the 'other groups' section. I think that is correct?

So in my other PR, I always make sure the field_mode value is passed in the configuration.

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 13]
```diff
@@ -91,6 +117,11 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
       }
     }
     else {
+
+      if ($this->configuration['handler_settings']['field_mode'] != 'admin') {
```
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: Similar to above, i think I have the logic working in https://github.com/amitaibu/og/pull/21 so we will want to either use that, or this.

File: src/services/OgSelectionManager.php [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+ * @see \Drupal\Core\Entity\Annotation\EntityReferenceSelection
+ * @see \Drupal\Core\Entity\EntityReferenceSelection\SelectionInterface
+ * @see plugin_api
+ */
+class OgSelectionManager extends SelectionPluginManager  {
+
+  public function getSelectionHandler(FieldDefinitionInterface $field_definition, EntityInterface $entity = NULL) {
+    $parent = parent::getSelectionHandler($field_definition, $entity);
+
+    if (!Og::isAudienceField($field_definition->getFieldStorageDefinition()->getName())) {
```
Comment (damiankloip) [inline src/services/OgSelectionManager.php]: See earlier comment, we already have a method for this we can use.

File: src/services/OgSelectionManager.php [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+      return $parent;
+    }
+
+    $options = array(
```
Comment (damiankloip) [inline src/services/OgSelectionManager.php]: nit: new array syntax can be used.

File: tests/src/Kernel/Entity/SelectionHandlerTest.php [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+   * We need to verify that the manager selection handler will use the default
+   * selection manager of the entity which the audience field referencing to.
+   *
+   * i.e: When the field referencing to node, we need verify we got the default
+   * node selection handler.
+   *
+   * @param mixed $match
+   *   The input text to be checked.
+   *
+   * @dataProvider providerTestCases
```
Comment (damiankloip) [inline tests/src/Kernel/Entity/SelectionHandlerTest.php]: I don't think a data provider is needed here, it makes the test harder to read, and there are not multiple cases.

File: tests/src/Kernel/Entity/SelectionHandlerTest.php [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\og\Og;
+use Drupal\user\Entity\User;
+
+/**
+ * Tests entity reference selection plugins.
+ *
+ * @group og
+ */
+class SelectionHandlerTest extends KernelTestBase {
```
Comment (damiankloip) [inline tests/src/Kernel/Entity/SelectionHandlerTest.php:23]: This class doesn't test much yet, I guess we are expanding this in a follow up PR?

File: src/services/OgSelectionManager.php [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+use Drupal\og\Og;
+
+/**
+ * Plugin type manager for Entity Reference Selection plugins.
+ *
+ * @see \Drupal\Core\Entity\Annotation\EntityReferenceSelection
+ * @see \Drupal\Core\Entity\EntityReferenceSelection\SelectionInterface
+ * @see plugin_api
+ */
+class OgSelectionManager extends SelectionPluginManager  {
```
Comment (damiankloip) [inline src/services/OgSelectionManager.php]: Not sure we really need this service? This could easily just be a method on the `OgSelectionHandler`? This is kind of just something it uses internally?

**Comment by RoySegall** [comment]
@damiankloip Any idea about the error i'm getting?

**Comment by mradcliffe** [comment]
@RoySegall, do you need to unit test the full stack or can you use UnitTestCase instead? With some mocks to getDefinition(s) it should be possible to instantiate a plugin.

It seems that the test is asserting output based on class and getConfiguration method. That seems simple enough for a UnitTestCase where you can pass in a mocked config entity.

**Comment by damiankloip** [comment]
Ideally we want both I would say. Let's get these tests working here and
follow up the rest.

**Comment by RoySegall** [comment]
I'm trying to debug this one and i'm able to do this only through simpletest but the error is the same.

The big question here is why when i'm using `$this->createEntityReferenceField('node', $group_content_type, OG_AUDIENCE_FIELD, $this->randomString(), 'node');` the tests are OK. But when i'm using `Og::createField(OG_AUDIENCE_FIELD, 'node', $group_content_type);` I get the error.

After some long debugging(and still going) I can point on the difference - the method `FieldStorageConfig::getSchema()` is being invoked in the first code and invoked in the second. I guess once I'll found out why I could fix the issue.

Any idea?

**Comment by damiankloip** [comment]
I think this is because we need some schema definitions for our custom field type. On a mobile at the moment,  I'll post an example when I'm at the computer.

**Comment by RoySegall** [comment]
But even after removing `OgMembershipReferenceItem::schema()` nothing it's still fails.

**Comment by damiankloip** [comment]
Not there. The config schemas.

**Comment by RoySegall** [comment]
Cool. I'm waiting for that. In the mean while I pushed the simple test as a reference. Once the tests will pass I'll remove it.

File: src/Og.php (hunk 5 above)
Comment (RoySegall) [inline src/Og.php]: For now it's comment out since this api got removed. Once the tests are passing I'll fix this one.

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 11 above)
Comment (RoySegall) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: I think `getSelectionHandler` since there will only one and no need for `Base` in the name.

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 10 above)
Comment (RoySegall) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: Don't think we need it. The selection handler will use the current user by default.

File: src/services/OgSelectionManager.php (hunk 18 above)
Comment (RoySegall) [inline src/services/OgSelectionManager.php]: We might want to inject it in the future. let's keep it like this and remove it in the future once we'll see the injection is redundant.

File: tests/src/Kernel/Entity/SelectionHandlerTest.php (hunk 17 above)
Comment (RoySegall) [inline tests/src/Kernel/Entity/SelectionHandlerTest.php:23]: Indeed. Once this test will pass we can go on with more complex tests.

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 11 above)
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: Just seems something like base in the name is more indicative of what's happening, as we're already in a selection handler.

File: .travis.yml [hunk 19]
```diff
@@ -1,7 +1,9 @@
 language: php
 sudo: false
 
-php: 5.5
+php:
+  - 5.7
```
Comment (damiankloip) [inline .travis.yml]: 5.5?

File: src/services/OgSelectionManager.php (hunk 18 above)
Comment (damiankloip) [inline src/services/OgSelectionManager.php]: If we are keeping it as a service, it shouldn't be in the services namespace, it also shouldn't really extend the SelectionPluginManager. It is not a plugin manager, it just uses the plugin manager.

**Comment by damiankloip** [comment]
@RoySegall I think https://github.com/amitaibu/og/pull/23 might help with the config schema exception in the test.

**Comment by RoySegall** [comment]
@damiankloip Please remind me to get you a beer at the next DrupalCon. The PR fixed the test. All that is left is minor fixes and we can merge it.

**Comment by damiankloip** [comment]
Beer is good, and I never forgot about it - don't worry. I would have rather the other PR was merged though so we don't lose that history! :)

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 12 above)
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: > Checking for empty() like this on a setting is asking for more bugs, I think we should keep explicit checking for types, if it's not set for whatever reason, we want to know about it

yeah, fully agree. no baby sitting code. Anyway, as discussed the field mode should be moved outside of the field config.

File: tests/src/Kernel/Entity/SelectionHandlerTest.php (hunk 16 above)
Comment (RoySegall) [inline tests/src/Kernel/Entity/SelectionHandlerTest.php]: I took it from `SelectionHandlerTest` and the tests work with this line.

**Comment by RoySegall** [comment]
@amitaibu Can you review please? 10x.

File: src/Og.php [hunk 20]
```diff
@@ -57,11 +59,25 @@ public static function createField($field_name, $entity_type, $bundle) {
       static::invalidateCache();
     }
 
+    $form_display_storage = \Drupal::entityManager()->getStorage('entity_form_display');
```
Comment (amitaibu) [inline src/Og.php:61]: I'd prefer to PR be only about the selection handler, not the form.

Comment (amitaibu) [inline src/Og.php:61]: (unless this is needed for PHP to not WSOD)

File: src/Og.php [hunk 21]
```diff
@@ -273,4 +289,34 @@ protected static function fieldInfo($field_name = NULL) {
...
+   *   The field name.
+   *
+   * @return OgSelection
+   * @throws \Exception
+   */
+  public static function getSelectionHandler($entity, $bundle, $field_name) {
+    $field_definition = FieldConfig::loadByName($entity, $bundle, $field_name);
+
+    if (!Og::isGroupAudienceField($field_definition)) {
+      throw new \Exception(t('The field @name is not an audience field.', ['@name' => $field_name]));
```
Comment (amitaibu) [inline src/Og.php]: t => format_string()

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 22]
```diff
@@ -62,7 +91,9 @@ public function setAccount(AccountInterface $account) {
    *   it.
    */
   protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS') {
-    $query = parent::buildEntityQuery($match, $match_operator);
+
+    $handler = $this->getSelectionHandler();
```
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: worth a comment.

File: tests/src/Kernel/Entity/SelectionHandlerTest.php [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+
+    // Create a group.
+    NodeType::create([
+      'type' => $group_type = Unicode::strtolower($this->randomMachineName()),
+      'name' => $this->randomString(),
+    ])->save();
+
+    // Create a group content type.
+    NodeType::create([
+      'type' => $group_content_type = Unicode::strtolower($this->randomMachineName()),
```
Comment (amitaibu) [inline tests/src/Kernel/Entity/SelectionHandlerTest.php]: move variable (e..g $group_content_type) before the create - to make more readable.

File: tests/src/Kernel/Entity/SelectionHandlerTest.php (hunk 16 above)
Comment (amitaibu) [inline tests/src/Kernel/Entity/SelectionHandlerTest.php]: I agree with @damiankloip here that it would be easier to follow the tests if we'll simple hardcode the value, for example

``` php
$this->assertEquals(get_class($this->selectionHandler->getSelectionHandler()), 'Drupal\node\Plugin\EntityReferenceSelection\NodeSelection');
```

File: src/Og.php (hunk 20 above)
Comment (RoySegall) [inline src/Og.php:61]: Not needed but could used this as a reference.

Comment (amitaibu) [inline src/Og.php:61]: ok.

File: tests/src/Kernel/Entity/SelectionHandlerTest.php (hunk 23 above)
Comment (RoySegall) [inline tests/src/Kernel/Entity/SelectionHandlerTest.php]: This is how it's done in core but there is an improvement i'd like to do: move part of the logic for setting up groups into traits so we could use this in other tests classes.

File: config/schema/og.schema.yml [hunk 24]
```diff
@@ -1,3 +1,22 @@
+field.storage_settings.og_membership_reference:
```
Comment (damiankloip) [inline config/schema/og.schema.yml:1]: This stuff already got merged.

File: tests/src/Kernel/Entity/SelectionHandlerTest.php (hunk 23 above)
Comment (amitaibu) [inline tests/src/Kernel/Entity/SelectionHandlerTest.php]: > move part of the logic for setting up groups into traits

Ok, but not in this PR please :)

**Comment by RoySegall** [comment]
@amitaibu Please check again. 10x.

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 25]
```diff
@@ -42,9 +41,39 @@ public function getAccount() {
...
+   *
+   * @param $key
+   *   Specific key from the configuration. Optional.
+   *
+   * @return array|string
+   *   Value of a configuration or all the configurations.
+   */
+  public function getConfiguration($key = NULL) {
+
+    if (!$key) {
```
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: `!isset()`

Apart from that, this is looking good now. Nice!

Comment (RoySegall) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: Don't think we need this. By default the key is NULL and if no value was passed `!$key` will return `TRUE` and the method will return all the configurations.

Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php]: For me it just makes for code that is less prone to bugs. `!$key` is just
any falsey value, which is not necessarily NULL.

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 7 above)
Comment (pfrenssen) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:26]: Is this intentional? If I look in a similar class like `\Drupal\taxonomy\Plugin\EntityReferenceSelection\TermSelection` then they also use `default:taxonomy_term` and not `taxonomy_term:default`.

**Comment by RoySegall** [comment]
@amitaibu This one ready for review. This PR test the results of the selection handler.

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 7 above)
Comment (RoySegall) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:26]: The idea is that we will take the original query object provided by the original selection handler(could be for node, term, user or even commerce product) and apply OG additional logic.

Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:26]: This change is :+1: from me :)

File: og.module [hunk 26]
```diff
@@ -102,7 +101,7 @@ function og_entity_insert(EntityInterface $entity) {
     return;
   }
 
-  if (!empty($entity->hasField('uid'))) {
+  if ($entity->hasField('uid')) {
```
Comment (amitaibu) [inline og.module]:
``` php
if ($entity instanceof EntityOwnerInterface) {
  // ...
  $uid = $entity->getOwnerId();
}
```

File: src/Og.php [hunk 27]
```diff
@@ -273,4 +288,36 @@ protected static function fieldInfo($field_name = NULL) {
...
+   *   The bundle name.
+   * @param $field_name
+   *   The field name.
+   * @param array $options
+   *   Overriding the default options of the selection handler.
+   *
+   * @return OgSelection
+   * @throws \Exception
+   */
+  public static function getSelectionHandler($entity, $bundle, $field_name, array $options = []) {
```
Comment (amitaibu) [inline src/Og.php]: I don't think we need an $option, just `$field_mode` or even `$is_admin = FALSE`, as In the OG complex widget we'll get the right selection handler without mocking the `handler_settings`

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 7 above)
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:26]: @RoySegall worth adding a comment to prevent people wondering.

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 28]
```diff
@@ -31,7 +30,7 @@ class OgSelection extends DefaultSelection {
    * @return AccountInterface
    */
   public function getAccount() {
-    if (empty($this->account)) {
+    if (empty($this->currentUser)) {
```
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:40]: @damiankloip pointed out we could maybe just mock the current user, instead of having this method. Anyway, not a must for this PR.

File: src/Plugin/EntityReferenceSelection/OgSelection.php [hunk 29]
```diff
@@ -42,9 +41,39 @@ public function getAccount() {
    * Set the current object account.
    *
    * @param AccountInterface $account
+   *   The user object.
+   * @return $this
```
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:52]: this looks wrong ;)

File: tests/src/Kernel/Entity/SelectionHandlerTest.php [hunk 30]
```diff
@@ -0,0 +1,179 @@
...
+   * @param $amount
+   *   The number of groups to create.
+   * @param User $user
+   *   The user object which own the groups.
+   * @param Bool $return_ids
+   *   Determine if the method will return the IDs or the group objects.
+   *
+   * @return ContentEntityBase[]
+   */
+  private function createGroups($amount, User $user, $return_ids = TRUE) {
```
Comment (amitaibu) [inline tests/src/Kernel/Entity/SelectionHandlerTest.php]: is $return_ids really needed?

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 28 above)
Comment (damiankloip) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:40]: I think I may be changing that in the other PR already. So yeah, leaving
here is ok.

**Comment by amitaibu** [comment]
I'm going to test this now, and fix the minor comments.

File: src/Plugin/EntityReferenceSelection/OgSelection.php (hunk 29 above)
Comment (amitaibu) [inline src/Plugin/EntityReferenceSelection/OgSelection.php:52]: this method is going soon, so I'm not bothering with it.

**Comment by amitaibu** [comment]
Merged, thanks!

https://github.com/amitaibu/og/commit/a704c6b11007c2c57869ee8e57b1e65fec72cef4

**Comment by damiankloip** [comment]
Squashing this makes it way harder to just revert the "Rename field_mode to is_admin" commit :)

**Comment by amitaibu** [comment]
Yeah, that's my default thing to do. But it throw at me an error I didn't know, so I gave up ;)

**Comment by damiankloip** [comment]
That's because that commit doesn't exist anymore because it was squashed :)

**Comment by amitaibu** [comment]
No, I meant prior to the squashing.

**Comment by damiankloip** [comment]
Oh :D Then I am confused also...
